### PR TITLE
mgr/dashboard_v2: Remove unnecessary meta classes

### PR DESCRIPTION
**Pull request by p-se**
Both commits will remove meta classes without substitution, as there's no substitution necessary, resulting in simpler code with fewer lines.

_First Commit_

Remove meta class `BaseControllerMeta` without substitution. The class is not required and just adds complexity to the code. Due to the removal of this class, it's not necessary to use the `six` module (in that file) anymore.

_Second Commit_

The second commit will simplify the `CephService` class by removing the meta class `ServiceMeta` together with the `Service` class and introducing `CephService` as a `CephServiceMixin` class.

In addition to the removal of these two classes, the injection of the mgr instance becomes unnecessary as all controller classes do have their own reference to the mgr object.

Furthermore, this commit will introduce a Python stub file which enables IDEs to correctly detect the type of the `mgr` instance within the newly created mixin.

Signed-off-by: Patrick Nawracay <pnawracay@suse.com>

### CODE REVIEW

**Comment by p-se** [comment]
Maybe squashed with 3beada2ac603a69f41cd9806db785d0633c37c69?

**Review by LenzGr — DISMISSED** [review]

**Comment by rjfd** [comment]
@p-na could you also make the same change to the `Service` class in `services/__init__.py`?

**Comment by votdev** [comment]
What is the reason for that?

**Comment by sebastian-philipp** [comment]
@votdev KISS. meta classes add a lot of complexity. They should only be used, if really necessary.

**Comment by p-se** [comment]
@votdev The reason is that using a meta class at this point makes the code unnecessarily complex and because there's no reason to use a meta class, I have removed it.

**Comment by votdev** [comment]
We can start the discussion about this again we already had when i introduced this. I personally do not like this coding style, i want to define a clear class interface. But if i'm the only one here i can not do anything against this PR. Because of this i can not approve this PR.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.pyi
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.pyi:4]: I would remove this file. Although I understand the reason why you are creating it, it feels strange that an interface only declares one attribute, and does not also declare all methods that are part of `CephServiceMixin` implementation. Instead of creating this file, I would add a class comment to the `CephServiceMixin` explaining that this mixin should be used with controller classes.

But if you decide to still stick with this file, could you please add above the first line:

```
# -*- coding: utf-8 -*-
from __future__ import absolute_import
```

**Review by rjfd — DISMISSED** [review]
After seeing the implementation, I'm starting to like the approach.
LGTM modulo the `.pyi` file comment

**Comment by rjfd** [comment]
@p-na please sync the merge of this PR with @s0nea, since PR #116 is already using the `CephService` class. I would recommend that PR #116 is merged first, and then you adapt this PR.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.pyi
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.pyi:4]: It makes PyCharm understand the mgr property. I personally could live without it, though

File: src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/tests/test_tcmu_iscsi.py]: I think we can remove this line completely.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/tools.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/tools.py]:
```python
mgr = None  # type: Module
´´´

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Comment by rjfd** [comment]
@p-na forgot to ask one thing. Could you add a commit with an update to the development documentation explaining how to use this mixin with an example?

File: src/pybind/mgr/dashboard_v2/services/ceph_service.pyi
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.pyi:4]: As this is the best way I found to make IDEs aware of the type of the `mgr` attribute within the `CephServiceMixin`, I'd love to keep that file. It's not meant to be used as a complete interface to `CephServiceMixin`.

**Review by sebastian-philipp — DISMISSED** [review]

**Comment by sebastian-philipp** [comment]
This needs to be rebased now.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.pyi
Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.pyi:4]: I continue to not like to have this file. And with the latest additions I'm even more against it. This will be a maintenance burden, and someone will eventually forget to add the necessary info here.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:139]: I don't see this mixin being used anywhere. Also, the function exposed by this mixin is particular to the rbd-mirroring component, and therefore it should be called `CephRbdMirrorMixin` and expose all functions related to rbd-mirroring.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:151]: The name is misleading. `CephPoolMixin.get_pools` doesn't return pools usable for exporting pools in the API.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:149]: This mixin class should be merged with the mixin class above, and renamed to `CephRbdMirrorMixin` as suggested in my previous comment.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:42]: The name is misleading in the context of `ceph_service.py`: It doesn't return generic daemons, instead it is tightly bound to the rbd-mirroring controller

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:77]: The name is misleading: `_get_daemon_health` doesn't return anything that is related to "health" of a generic daemon. Instead, it is tightly bound to the rbd-mirroring controller.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:221]: the functions in these mixin classes shouldn't use the `@ViewCache` decorator. The `@ViewCache` decorator should be use solely in controller classes so that it's explicit in the controller code which data is being cached, otherwise when I want to re-use this code I may not now that it's already being cached, and worst, I might not want it to be cached.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:139]: The name is misleading: It doesn't providing generic daemon services. in stead it is bound to the mirroring controller.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:109]: I don't see the benefit of moving this code outside the `CephServiceMixin` class. I don't see a problem of calling other mixins functions inside a mixin.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:222]: Again. misleading.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:151]: why is this method public? is it being used anywhere?

**Review by rjfd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/services/ceph_service.py:1]: I don't like the way, you spread out `get_daemons_and_pools`. To me, your refactorization does not improve the readability, instead it makes it hard to identify that all functions and methods are only usable together and are tightly coupled to each other.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.py:151]: Any suggestions?

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Comment by sebastian-philipp** [comment]
I'd recommend to split this PR in two: 

* the first one containing everything except your `get_dameons_and_pools` refactorization and
*  the second one containing just the `get_dameons_and_pools` refactorization

**Comment by rjfd** [comment]
@p-na  I agree with @sebastian-philipp suggestion of splitting this PR in two.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.py:221]: Does that completely exclude Mixins to be decorated by the `@ViewCache` decorator?

Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.py:221]: only in the mixin function declaration (mixin file). These mixin functions can, and some should, be used in controllers' functions decorated with `@ViewCache`.

File: src/pybind/mgr/dashboard_v2/services/ceph_service.py
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.py:109]: > I don't see a problem of calling other mixins functions inside a mixin.

There's a problem with the `mgr` instance, which is not yet there when calling `CephServiceMixin.<method>` from `CephRbdMirrorMixin`, for instance. Do you have a better solution for this, instead of outsourcing the methods to the module like I did?

Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.py:109]: Here's a good case where mixins fail to provide the necessary infrastructure. The idea of these "services" classes is to implement the "business" logic common to several controllers. It should be natural for services to call each other in a compositional way. I should be able to create a more complex service by composition of functions of existing services.

With the current mixin solution, services logic/implementation is not independent of controllers because they require their `mgr` instance. To solve your particular problem, I would do this:

```
CephServiceMixin.mgr = cls.mgr
return CephServiceMixin.get_service_map(service_name)
```

But it is still ugly though :(

Having functions declared global in the module scope, will fail to create an organized structure of the code, unless we add big comment notes separating each set of functions according to the mixin class they primary belong.

Comment (p-se) [reply src/pybind/mgr/dashboard_v2/services/ceph_service.py:109]: This is the same problem we have with controllers, the `mgr` instance is injected to each controller (but it could be passed in the constructor!). Maybe we could do the same for the mixins. Would feel better than injecting the attribute right before the call.

**Comment by p-se** [comment]
Closed this PR due to changes which need to be ahead of these changes. The changes of this PR, alongside the ones from @rjfd will be part of two new PRs.

**Pull request closed without merging**
